src/daemon/mq.rs

```diff
@@ -57,96 +58,135 @@ pub enum QueueTask {
                 write!(f, "resource class removed for '{}' ", ca)
             }
-            QueueTask::UnexpectedKey { ca, rcn, .. } => {
+            Task::UnexpectedKey { ca, rcn, .. } => {
                 write!(f, "unexpected key found for '{}' resource class: '{}'", ca, rcn)
             }
         }
     }
 }
 
+//------------ TaskQueue ----------------------------------------------------
+
 #[derive(Debug)]
-pub struct MessageQueue {
-    q: RwLock<VecDeque<QueueTask>>,
+pub struct TaskQueue {
+    q: RwLock<PriorityQueue<Task, Priority>>,
 }
 
-impl Default for MessageQueue {
+impl Default for TaskQueue {
     fn default() -> Self {
-        let mut vec = VecDeque::new();
-        vec.push_back(QueueTask::ServerStarted);
-        MessageQueue { q: RwLock::new(vec) }
+        let mut q = PriorityQueue::new();
+        q.push(Task::ServerStarted, now());
+
+        TaskQueue { q: RwLock::new(q) }
     }
 }
 
-impl MessageQueue {
-    pub fn pop_all(&self) -> Vec<QueueTask> {
-        let mut res = vec![];
+impl TaskQueue {
+    pub fn pop(&self, due_before: Priority) -> Option<Task> {
         let mut q = self.q.write().unwrap();
-        while let Some(evt) = q.pop_front() {
-            res.push(evt);
+
+        let has_item = if let Some((_, priority)) = q.peek() {
+            priority > &due_before
+        } else {
+            false
+        };
+
+        if has_item {
+            q.pop().map(|(item, _)| item)
+        } else {
+            None
         }
-        res
     }
 
-    fn schedule(&self, task: QueueTask) {
+    /// Schedules a task for the given priority. If the equivalent task
+    /// was already present, then it will get the highest of the two
+    /// priorities.
+    ///
+    /// Many tasks are planned with a high priority - e.g. if they are
+    /// triggered through CA events. Other tasks may be planned for the
+    /// future (e.g. sync with parent tomorrow). The latter can be moved
+    /// forward when circumstances dictate.
+    ///
+    /// Recurring tasks will typically be re-added by the Scheduler when
+    /// needed (and can then be moved forward if needed).
+    fn schedule(&self, task: Task, priority: Priority) {
         let mut q = self.q.write().unwrap();
-        q.push_back(task);
+
+        let prio_opt = q.get_priority(&task).copied();
+
+        match prio_opt {
+            None => {
+                debug!("Adding task: {}, with priority: {}", task, priority);
+                q.push(task, priority);
+            }
+            Some(existing_priority) => {
+                if existing_priority < priority {
+                    debug!(
+                        "Re-prioritising task: {} from: {} to: {}",
+                        task, existing_priority, priority
+                    );
+                    q.change_priority(&task, priority);
+                } else {
+                    debug!("Keeping existing task: {} with higher priority: {}", task, priority);
```

Comment:
or equal

Reply:
Yes, but it's an unlikely corner case that does not matter in this context so I opted for brevity.